drivers/eeprom/eeprom_emulator.c

```diff
@@ -0,0 +1,751 @@
+		}
+
+		data8 += sizeof(buf);
+		address += sizeof(buf);
+		len -= sizeof(buf);
+	}
+
+	if (len) {
+		/* unaligned end */
+		rc = eeprom_emu_word_read(dev, address, buf);
+		if (rc) {
+			goto end;
+		}
+
+		if (write_op) {
+			memcpy(buf, data8, len);
+			rc = eeprom_emu_word_write(dev, address, buf);
+		} else {
+			memcpy(data8, buf, len);
+		}
+
+	}
+
+end:
+	k_mutex_unlock(&dev_data->lock);
+	if (write_op) {
+		(void)flash_write_protection_set(dev_config->flash_dev, true);
+	}
+
+	return rc;
+}
+
+static int eeprom_emu_read(const struct device *dev, off_t address, void *data,
+			   size_t len)
+{
+	return eeprom_emu_rw(dev, address, data, len, false);
+}
+
+static int eeprom_emu_write(const struct device *dev, off_t address,
+			    const void *data, size_t len)
+{
+	return eeprom_emu_rw(dev, address, (void *)data, len, true);
+}
+
+static size_t eeprom_emu_size(const struct device *dev)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	return dev_config->size;
+}
+
+static int eeprom_emu_init(const struct device *dev)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	struct eeprom_emu_data *dev_data = DEV_DATA(dev);
+	off_t offset;
+	uint8_t buf[dev_config->flash_cbs];
+	int rc;
+
+	k_mutex_init(&dev_data->lock);
+	if (!device_is_ready(dev_config->flash_dev)) {
+		__ASSERT(0, "Could not get flash device binding");
+		return -ENODEV;
+	}
+
+	/* Find the page offset */
+	dev_data->page_offset = 0U;
+	dev_data->write_offset = dev_config->page_size - sizeof(buf);
+	while (dev_data->page_offset < dev_config->flash_size) {
+		offset = dev_data->page_offset + dev_data->write_offset;
+		rc = eeprom_emu_flash_read(dev, offset, buf, sizeof(buf));
+		if (rc) {
+			return rc;
+		}
+
+		if (!eeprom_emu_is_word_used(dev, buf)) {
+			break;
+		}
+
+		dev_data->page_offset += dev_config->page_size;
+	}
+
+	if (dev_data->page_offset == dev_config->flash_size) {
+		__ASSERT(0, "All pages are invalid, is this a EEPROM area?");
+		return -EINVAL;
+	}
+
+	dev_data->write_offset = dev_config->size;
+
+	/* Update the write offset */
+	while ((dev_data->write_offset + sizeof(buf)) < dev_config->page_size) {
+		offset = dev_data->page_offset + dev_data->write_offset;
+		rc = eeprom_emu_flash_read(dev, offset, buf, sizeof(buf));
+		if (rc) {
+			return rc;
+		}
+
+		if (!eeprom_emu_is_word_used(dev, buf)) {
+			break;
+		}
+
+		dev_data->write_offset += sizeof(buf);
+	}
+
+	/* dev_data->write_offset reaches last possible location, compaction
+	 * might have been interrupted: call eeprom_emu_compactor again, but
+	 * only in case we are using a write-enabled eeprom
+	 */
+	if ((!dev_config->readonly) &&
+	    ((dev_data->write_offset + sizeof(buf)) >= dev_config->page_size)) {
+		/* Need to write in next step, disable flash write protection */
+		if (flash_write_protection_set(dev_config->flash_dev, false)) {
+			__ASSERT(0, "cannot disable flash write protection");
+			return -EIO;
+		}
+
+		rc = eeprom_emu_compactor(dev);
```

Comment:
:question:
`rambuf` (if available) is not initialized yet, In that case this call to `eeprom_emu_compactor()` will write a garbage to the EEPROM.
Might be I wrong?

Reply:
You are right, I am going to rework `eeprom_emu_compactor()` so that it doesn't use `rambuf`.

Reply:
I think usage of valid `rambuf` is actually great idea, I appreciated that. Maybe you can just set the rambuf pointer to NULL for this call, and restore it once it return.

Reply:
The latest commit should resolve this.